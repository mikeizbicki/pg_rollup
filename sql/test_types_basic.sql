create or replace language plpython3u;
create extension if not exists pg_rollup;

create temporary table test (
    id serial primary key,
    name text,
    num int
);

insert into test (name,num) values
    ('alice', 1),
    ('alice', 2),
    ('alice', 3),
    ('alice', 4),
    ('alice', 5),
    ('bill', 5),
    ('bill', 5),
    ('bill', 5),
    ('charlie', 1),
    ('charlie', 1),
    ('charlie', 1),
    ('charlie', 3),
    ('charlie', NULL),
    ('dave', 4),
    ('elliot', 5),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 8),
    (NULL, 9),
    (NULL, NULL),
    (NULL, NULL),
    (NULL, NULL),
    (NULL, NULL);

select create_rollup(
    'test',
    'test_rollup1',
    wheres => 'name'
);

select create_rollup(
    'test',
    'test_rollup2',
    wheres => 'name,num'
);

select create_rollup(
    'test',
    'test_rollup3',
    wheres => 'name',
    rollups => $$
        sum(num) as sum,
        count(*) as count_all,
        count(num),
        max(num),
        min(num)
    $$
);

select create_rollup(
    'test',
    'test_rollup4',
    rollups => $$
        sum(num) as sum,
        count(*) as count_all,
        count(num),
        max(num),
        min(num)
    $$
);

select assert_rollup('test_rollup1');
select assert_rollup('test_rollup2');
select assert_rollup('test_rollup3');
select assert_rollup('test_rollup4');

insert into test (name,num) values
    ('alice', 1),
    ('alice', 2),
    ('alice', 3),
    ('alice', 4),
    ('alice', 5),
    ('bill', 5),
    ('bill', 5),
    ('bill', 5),
    ('charlie', 1),
    ('charlie', 1),
    ('charlie', 1),
    ('charlie', 3),
    ('charlie', NULL),
    ('dave', 4),
    ('elliot', 5),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 8),
    (NULL, 9),
    (NULL, NULL),
    (NULL, NULL),
    (NULL, NULL),
    (NULL, NULL);


select assert_rollup('test_rollup1');
select assert_rollup('test_rollup2');
select assert_rollup('test_rollup3');
select assert_rollup('test_rollup4');
