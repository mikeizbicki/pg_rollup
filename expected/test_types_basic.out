SET client_min_messages TO WARNING;
create table test (
    id serial primary key,
    name text,
    num int
);
insert into test (name,num) values
    ('alice', 1),
    ('alice', 2),
    ('alice', 3),
    ('alice', 4),
    ('alice', 5),
    ('bill', 5),
    ('bill', 5),
    ('bill', 5),
    ('charlie', 1),
    ('charlie', 1),
    ('charlie', 1),
    ('charlie', 3),
    ('charlie', NULL),
    ('dave', 4),
    ('elliot', 5),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 8),
    (NULL, 9),
    (NULL, NULL),
    (NULL, NULL),
    (NULL, NULL),
    (NULL, NULL);
create materialized view test_rollup1 as (
    select count(*),name
    from test
    group by name
);
create materialized view test_rollup2 as (
    select count(*),name,num
    from test
    group by name,num
);
create materialized view test_rollup3 as (
    select
        name,
        sum(num) as sum,
        count(*) as count_all,
        count(num) count_num,
        max(num),
        min(num)
    from test
    group by name
);
create materialized view test_rollup4 as (
    select 
        sum(num) as sum,
        count(*) as count_all,
        count(num),
        max(num),
        min(num)
    from test
);
select assert_rollup('test_rollup1');
 assert_rollup 
---------------
 
(1 row)

select assert_rollup('test_rollup2');
 assert_rollup 
---------------
 
(1 row)

select assert_rollup('test_rollup3');
 assert_rollup 
---------------
 
(1 row)

select assert_rollup('test_rollup4');
 assert_rollup 
---------------
 
(1 row)

insert into test (name,num) values
    ('alice', 1),
    ('alice', 2),
    ('alice', 3),
    ('alice', 4),
    ('alice', 5),
    ('bill', 5),
    ('bill', 5),
    ('bill', 5),
    ('charlie', 1),
    ('charlie', 1),
    ('charlie', 1),
    ('charlie', 3),
    ('charlie', NULL),
    ('dave', 4),
    ('elliot', 5),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 8),
    (NULL, 9),
    (NULL, NULL),
    (NULL, NULL),
    (NULL, NULL),
    (NULL, NULL);
select assert_rollup('test_rollup1');
 assert_rollup 
---------------
 
(1 row)

select assert_rollup('test_rollup2');
 assert_rollup 
---------------
 
(1 row)

select assert_rollup('test_rollup3');
 assert_rollup 
---------------
 
(1 row)

select assert_rollup('test_rollup4');
 assert_rollup 
---------------
 
(1 row)

drop table test cascade;
